### PR TITLE
[WIP] Add `METEORScore`

### LINKED ISSUE #595: Add METEOR metric
## 🚀 Feature

Add another NLP metric - **METEOR** ([Lavie and Agarval, 2007](https://www.cs.cmu.edu/~alavie/METEOR/pdf/Lavie-Agarwal-2007-METEOR.pdf)).

### Motivation

METEOR is another metric used for the machine translation evaluation similarly to BLEU, however, it demonstrates a higher correlation with human judgements of translation quality.

### Pitch

To support the METEOR metric will be likely to require `nltk` dependency. The `nltk` package is also, however, needed for the ROUGE metric. METEOR thus should not bring any new dependency.

### Additional context

I will be happy to start working on this feature next week if desired to be added :)

### PR DESCRIPTION
**Pull request by stancld**
## What does this PR do?

This PR is currently heavilly WIP. 
This PR adds a new NLP metric `METEORScore` according to version 1.5 [Meteor Universal: Language Specific Translation Evaluation for Any Target Language](https://aclanthology.org/W14-3348.pdf) by Michael Denkowski and Alon Lavie. The documentation for the current official implementation in Java (see [Meteor 1.5: Automatic Machine Translation Evaluation System](https://www.cs.cmu.edu/~alavie/METEOR/README.html)).

The new intended Torchmetrics implementation should be furthermore compatible with the initial version - [Meteor: An Automatic Metric for MT Evaluation with High Levels of Correlation with Human Judgments](https://aclanthology.org/W07-0734.pdf) by Alon Lavie and Abhaya Agarwal - which is implemented within the [`nltk` package](https://github.com/nltk/nltk/blob/develop/nltk/translate/meteor_score.py).

## Some implementation details:
- This new intended implementation relies on `SnowStemmer` and `Wordnet` directly available from the `nltk` package (hopefully compatible with 1.5 Java implementation)
- It uses the list of function words available from [Meteor 1.5: Automatic Machine Translation Evaluation System](https://www.cs.cmu.edu/~alavie/METEOR/README.html) (Question: Are we able to comply with the LICENSE?)
- TBA ...

## TODOs:
- [ ] Finish docs
- [ ] Implemente weights for individual matches coming from different match stages.
- [ ] Find out the way for the usage of paraphrase tables.
- [ ] Write tests for METEOR v1.5
- [ ] Ensure that all supporting data can be used under their LICENSE
- [ ] Fix tests for the configuration of `ddp=True & dist_sync_on_step=False`

Fixes #595

## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## PR review

**Not ready for review yet** :]

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: torchmetrics/text/meteor.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+            dist_sync_on_step=dist_sync_on_step,
+            process_group=process_group,
+            dist_sync_fn=dist_sync_fn,
+        )
+
+        self.stemmer = _NLTKStemmerWrapper(stemmer)
+        self.wordnet = _NLTKWordnetWrapper(wordnet)
+        self.alpha = alpha
+        self.beta = beta
+        self.gamma = gamma
```
Comment (SkafteNicki) [inline torchmetrics/text/meteor.py:102]: I assume we need to do some validation of these parameters

**Review by SkafteNicki — COMMENTED** [review]
Remember to insert a reference in the docs to both the functional and modular implementation

File: torchmetrics/text/meteor.py (hunk 1 above)
Comment (stancld) [reply torchmetrics/text/meteor.py:102]: Yes, `gamma` should be also between 0 and 1. Not sure about `beta`. I'll have a look :]

**Comment by stancld** [comment]
@Borda @SkafteNicki Hello, could I please kindly ask you if you have any idea if there is a good way how to handle such a conditional import like this one below?

```python
class _NLTKStemmerWrapper:
    """TorchMetrics wrapper for `nltk` stemmers."""

    if not _NLTK_AVAILABLE:
        raise ValueError(
            "Stemmer requires that nltk is installed. Use `pip install nltk` or or `pip install torchmetrics[text]."
        )
    from nltk.stem import PorterStemmer, StemmerI

    _STEMMER_CLASS: Dict[str, StemmerI] = {"porter": PorterStemmer}
```

Since the `nltk` package is a part of the `text` requirements only, it will raise a `ValueError` during tests as of now. On the other hand, if I move this conditional import within an `__init__` method, I face an error during the `pickle.loads`.

Thanks a lot in advance O:]

**Comment by SkafteNicki** [comment]
Hi @stancld, I would just have this at the beginning of the file:
```python
if _NLTK_AVAILABLE:
    # import everything from nltk you need
    from nltk.stem import PorterStemmer, StemmerI
else:
    # set them to None in this case
    PorterStemmer, Stemmerl = None, None
```
To my understanding `_NLTKStemmerWrapper` and the other wrapper should not be called by the user and therefore there is no need for do the value check on them when it is already done by both the functional and modular implementation before the wrappers are initialized.

**Comment by Borda** [comment]
@stancld still a draft? :rabbit:

**Comment by stancld** [comment]
METEORScore is heavy dependent on some external data.. I'll need to think about that a little bit more, if we're allowed to use them, and eventually how.

**Comment by Borda** [comment]
@stancld hi have you come with some solution or ou may write here what options you're thinking about and we may help...

**Comment by stancld** [comment]
@Borda - Hi, sorry for being silent on this PR. The problem with the official METEORScore (it should be the one in version `1.5`, not `1.0` also implemented in `nltk`) is that it relies on a few external sources:
- a list of function words,
- a paraphrase table.

I guess the official documentation for the Java implementation should be [here](https://www.cs.cmu.edu/~alavie/METEOR/README.html).

With that being said, I'm not sure if we're allowed to use these external data and if so, what's the best way how to store them for our implementation? Do you have any thoughts on this? Thanks a lot :]

**Comment by Borda** [comment]
> With that being said, I'm not sure if we're allowed to use these external data and if so, what's the best way how to store them for our implementation? Do you have any thoughts on this? Thanks a lot :]

@lantiga, could you pls check/provide your feedback?

**Comment by Borda** [comment]
let's close it for now... :flags:

**Pull request closed without merging**